Docs 1826 restart impaired vs unresponsive workers hn (#234)
mulesoft/docs-runtime-manager: modules/ROOT/pages/cloudhub-impaired-worker.adoc

= Impaired Worker Monitoring 
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

image:logo-cloud-active.png[link="/runtime-manager/deployment-strategies", title="CloudHub"]
image:logo-hybrid-disabled.png[link="/runtime-manager/deployment-strategies", title="Hybrid"]
image:logo-server-disabled.png[link="/runtime-manager/deployment-strategies", title="Anypoint Platform PCE"]
image:logo-rtf-disabled.png[link="/runtime-fabric/deploy-to-runtime-fabric", title="Runtime Fabric"]

CloudHub continuously performs health checks on the underlying virtual machines where Mule applications are running. 
When a worker is impaired, CloudHub automatically attempts to restart the worker and relaunch applications.

A worker is impaired when there is a hardware or software failure that negatively impacts the running application deployed on the worker.
Hardware and software failures include network unavailability, memory issues, server hardware failures, and file corruption.

> When CloudHub restarts the worker:

Reviewer: When CloudHub restarts the impaired worker @hannanelson

* If a retry is successful, the underlying worker and app restart without any further action.
+
It might take up to five minutes for a worker to restart.
* If the application fails to start after five attempts, the status of the application is *Failed* or *Unrecoverable Runtime Error*.
+
In this case, you must manually restart the application.

The logs indicate when a worker is restarted automatically and any failures.

[NOTE]
The automatic worker-restart functionality is different from the automatic application-restart functionality, which is configurable using the *Automatically restart application when not responding* option on the *Settings* page for the app. 
CloudHub automatically attempts to restart impaired workers, regardless of the app setting. 

== Scheduled Maintenance Activities

CloudHub checks for scheduled maintenance activities for worker instances that cause potential downtime for the running applications: reboot, stop, start, or retirement of the instances.

> A notification appears in the top right of the *Settings* page for the application page, notifying you of upcoming maintenance activities:

Reviewer: A notification appears in the top right of the *Settings* page for the application and application list page, notifying you of upcoming maintenance activities:

.The arrow shows the notification icon in the *Settings* page.
image::cloudhub-notification-icon.png[Notification icon in the *Settings* page]

Click the icon to display the notification, which includes the date that CloudHub will automatically restart the worker.

> You can restart your apps manually or wait for CloudHub to restart them automatically. 

Reviewer: You can restart your apps manually or wait for CloudHub to restart them automatically. A reminder notification will appear 1 week before the restart. 

If the platform performs the restart on your behalf, you will see the following message in the application log:
"Your application was running on an impaired worker instance. To avoid any downtime, the platform has restarted the application."
You will also see a notification indicating that the restart was done.

Reviewer: @hannanelson

> = Impaired Worker Monitoring 

Reviewer 2: lol

Reviewer 2: INSTRUCTIONS FOR EDIT

Can I get in line for a PR review when you return? Green only, as soon as you have time.
https://github.com/mulesoft/docs-runtime-manager/pull/234
Because of the graphics, it looks like more than it is - really just 3 topics affected plus nav.adoc. (edited) 
Cleaned up/clarified app monitoring vs worker monitoring. Also updated icons, removed PCF, added RTF.

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

image:logo-cloud-active.png[link="/runtime-manager/deployment-strategies", title="CloudHub"]
image:logo-hybrid-disabled.png[link="/runtime-manager/deployment-strategies", title="Hybrid"]
image:logo-server-disabled.png[link="/runtime-manager/deployment-strategies", title="Anypoint Platform PCE"]
image:logo-rtf-disabled.png[link="/runtime-fabric/deploy-to-runtime-fabric", title="Runtime Fabric"]

> CloudHub continuously performs health checks on the underlying virtual machines where Mule applications are running. 

Reviewer 2: Briefer alternative:
```suggestion
CloudHub continuously verifies the health of the virtual machines running Mule applications. 
```

> When a worker is impaired, CloudHub automatically attempts to restart the worker and relaunch applications.

Reviewer 2: This sentence equates "worker" with the VM; is that correct? Or is a "worker" just part of a VM, or something else? Do we need to clarify?

A worker is impaired when there is a hardware or software failure that negatively impacts the running application deployed on the worker.

> Hardware and software failures include network unavailability, memory issues, server hardware failures, and file corruption.

Reviewer 2: Minimalist alternative:

"If a worker is impaired by a system failure such as network unavailability, server hardware or memory issues, or file corruption, CloudHub automatically attempts to restart the worker and relaunch applications. If successful, the worker and app restart within five minutes, without any further action."

"However, if the worker restarts but the application fails to resume after five attempts, the application's log  status changes to...and you must manually restart it."

When CloudHub restarts the impaired worker:

* If a retry is successful, the underlying worker and app restart without any further action.
+
It might take up to five minutes for a worker to restart.
* If the application fails to start after five attempts, the status of the application is *Failed* or *Unrecoverable Runtime Error*.
+
In this case, you must manually restart the application.

> The logs indicate when a worker is restarted automatically and any failures.

Reviewer 2: See suggestion above.
Does the worker always restart, and the issue lie only with the application not restarting?
Is the text about the log necessary? Is that the only place that the application status change is noted?

Author: The Status appears in the apps list and in the logs. You might want to know that the logs show that an app was restarted without your intervention.

[NOTE]
The automatic worker-restart functionality is different from the automatic application-restart functionality, which is configurable using the *Automatically restart application when not responding* option on the *Settings* page for the app. 

> CloudHub automatically attempts to restart impaired workers, regardless of the app setting. 

Reviewer 2: This complicates the narrative flow further. So I have a new organizational suggestion:
1. If worker fails and causes app to fail, CloubHub uses auto restart on worker
2. Worker restarts (you can see in log)
3. If app is configured to auto restart and does, no further action (might take 5 min)
4. If app is not configured to auto restart or does not after 5 attempts, status changes (you can see in log)
5. You must manually restart app (and consider config auto if not?)

Author: Well, no. 3 and 4 don't apply. The app is restarted automatically when the worker restarts. Configurable auto-restart is different - and is the point of this bug!

== Scheduled Maintenance Activities

> CloudHub checks for scheduled maintenance activities for worker instances that cause potential downtime for the running applications: reboot, stop, start, or retirement of the instances.

Reviewer 2: transition and chrono reorg suggestion:

"In addition to automatically restarting failed workers, CloudHub also restarts workers after scheduled maintenance. One week before such maintenance, an icon appears <here>. Clicking the icon displays a notification that tells you when can expect Cloudhub to restart the worker and your apps."

QUESTIONS:
Does the worker and app continue to work during the maint?
Do they stop and then restart, or just restart?
Does the icon appear in the top right, or the notification; they are not the same thing, right?
Why does the user need to know about the restart: is there an up-to-5-min lull?
Why would the user want to manually restart: to avoid the lull, or only if the app is not configured to restart or fails?
Should we combine the texts about failure and maint, if the chronology is the same?
Does the message appear in both cases: failure and maint?
If a worker is retired, is this transparent: is there a replacement?

Author: No. Hence "downtime".
Restart.
The screenshot shows the icon. You click the icon to see the notification.
I guess so.
To schedule at their convenience.
No.
Only maint.
Yes.

> A notification appears in the top right of both the *Applications* page and the *Settings* page for the application, notifying you of upcoming maintenance activities:

Reviewer 2: notification or icon appears? "a notification icon"?

.The arrow shows the notification icon on the *Settings* page for the application.
image::cloudhub-notification-icon.png[Notification icon on the *Settings* page for the application]

Click the icon to display the notification, which includes the date that CloudHub will automatically restart the worker.

> You can restart your apps manually or wait for CloudHub to restart them automatically. A reminder notification appears one week before the automatic application restart.

Reviewer 2: Why would you restart manually? Do they not restart if not config to do so? In that case, CH cannot start them auto?
Why pair this with the notification text? Is there something that the user should do in the meantime?

Author: convenience. The config does not enter into this scenario.

> If CloudHub restarts the app, the following message appears in the application log:

Reviewer 2: Does CH restart the app, or does the app automatically restart because of how it is configured? 
Is this scenario (maint) different from failure?

Author: CH restarts the app. This is a restart of the worker, which is different than the auto-app restart. The app is restarted as a result of the worker restart.

[source,console]
----
Your application was running on an impaired worker instance. To avoid any downtime, the platform has restarted the application.
----

> You also see a notification indicating that the restart occurred.

Reviewer 2: This sounds repetitive. How is this notification of restart different from the message in the log?

Author: Notifications appear in the UI.
